agent/engine/dockerauth/ecr.go

```diff
@@ -17,54 +17,115 @@ import (
+	"github.com/aws/amazon-ecs-agent/agent/async"
 	"github.com/aws/amazon-ecs-agent/agent/ecr"
 	ecrapi "github.com/aws/amazon-ecs-agent/agent/ecr/model/ecr"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	"github.com/aws/aws-sdk-go/aws"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
 )
 
+type cacheKey struct {
+	region           string
+	roleARN          string
+	registryID       string
+	endpointOverride string
+}
+
 type ecrAuthProvider struct {
-	authData *api.ECRAuthData
-	client   ecr.ECRClient
+	tokenCache async.Cache
+	factory    ecr.ECRFactory
+	cacheLock  sync.RWMutex
 }
 
-const proxyEndpointScheme = "https://"
+const (
+	roundtripTimeout = 5 * time.Second
+	// MinimumJitterDuration is the minimum duration to mark the credentials
+	// as expired before it's actually expired
+	MinimumJitterDuration = 30 * time.Minute
+	proxyEndpointScheme   = "https://"
+)
+
+// String formats the cachKey as a string
+func (key *cacheKey) String() string {
+	return fmt.Sprintf("%s-%s-%s-%s", key.roleARN, key.region, key.registryID, key.endpointOverride)
+}
 
 // NewECRAuthProvider returns a DockerAuthProvider that can handle retrieve
 // credentials for pulling from Amazon EC2 Container Registry
-func NewECRAuthProvider(authData *api.ECRAuthData, clientFactory ecr.ECRFactory) DockerAuthProvider {
-	if authData == nil {
-		return &ecrAuthProvider{}
-	}
-	log.Tracef("Getting client in %s with endpoint %s", authData.Region, authData.EndpointOverride)
+func NewECRAuthProvider(ecrFactory ecr.ECRFactory, cache async.Cache) DockerAuthProvider {
 	return &ecrAuthProvider{
-		authData: authData,
-		client:   clientFactory.GetClient(authData.Region, authData.EndpointOverride),
+		tokenCache: cache,
+		factory:    ecrFactory,
 	}
 }
 
 // GetAuthconfig retrieves the correct auth configuration for the given repository
-func (authProvider *ecrAuthProvider) GetAuthconfig(image string) (docker.AuthConfiguration, error) {
-	if authProvider.authData == nil {
-		return docker.AuthConfiguration{}, fmt.Errorf("ecrAuthProvider cannot be used without AuthData")
+func (authProvider *ecrAuthProvider) GetAuthconfig(image string,
+	authData *api.ECRAuthData) (docker.AuthConfiguration, error) {
+	if authData == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("ecr auth: miissing container auth data")
+	}
+
+	// First try to get the token from cache, if the token does not exist,
+	// then call ECR api to get the new token
+	key := cacheKey{
+		region:           authData.Region,
+		endpointOverride: authData.EndpointOverride,
+		registryID:       authData.RegistryID,
+	}
+
+	// If the container is using execution role credentials to pull,
+	// add the roleARN as part of the cache key so that docker auth for
+	// containers pull with the same role can be cached
+	if !utils.ZeroOrNil(authData.GetPullCredentials()) {
+		key.roleARN = authData.GetPullCredentials().RoleArn
+	}
+	authProvider.cacheLock.RLock()
```

Comment:
By acquiring the lock here and deferring the unlock until the end of the function, you've now made the lock scope possibly include an API call.  You should reduce the scope so it only covers reads + mutations, and not off-box API calls.